### PR TITLE
Replace F extension with string interpolation

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

**Comment by pchote** [comment]
There are a bunch of places (e.g. the first several exception changes in the files view) where this is inserting arbitrary newlines that IMO hurt readability more than they help. Can we please unstage those?

**Comment by teinarss** [comment]
Updated.

**Comment by teinarss** [comment]
> 
> 
> Maybe reconsider this one - there is an option even better than what ReSharper implies. https://stackoverflow.com/a/63904747

The only places where we are using `{x:x2}` style is in none hot paths, as far as i can see, and this PR doesn't introduces any changes  with `valueType.ToString(format)` to `{valueType:format}`, which causes the boxing.

**Comment by dragunoff** [comment]
How would this work with localization? :thinking:

**Comment by pchote** [comment]
The translation library has its own string replacement system, so won't interact with this at all.

File: OpenRA.Game/Network/GameServer.cs [hunk 1]
```diff
@@ -200,9 +200,9 @@ public GameServer(MiniYaml yaml)
 						.FirstOrDefault(m => m.Id == Mod);
 
 					if (guessMod != null)
-						ModTitle = "{0}".F(guessMod.Title);
+						ModTitle = $"{guessMod.Title}";
```
Comment (pchote) [inline OpenRA.Game/Network/GameServer.cs]: Redundant format string.

File: OpenRA.Game/Game.cs [hunk 2]
```diff
@@ -976,16 +976,16 @@ public static void LoadMap(string launchMap)
 		{
 			var orders = new List<Order>
 			{
-				Order.Command("option gamespeed {0}".F("default")),
-				Order.Command("state {0}".F(Session.ClientState.Ready))
+				Order.Command($"option gamespeed {"default"}"),
```
Comment (pchote) [inline OpenRA.Game/Game.cs]: Redundant format string.

File: OpenRA.Game/Scripting/ScriptMemberExts.cs [hunk 3]
```diff
@@ -32,16 +32,16 @@ public static string LuaDocString(this Type t)
...
 			return ret;
 		}
 
 		public static string LuaDocString(this ParameterInfo pi)
 		{
-			var ret = "{0} {1}".F(pi.ParameterType.LuaDocString(), pi.Name);
+			var ret = $"{pi.ParameterType.LuaDocString()} {pi.Name}";
 			if (pi.IsOptional)
-				ret += " = {0}".F(pi.DefaultValue != null ? pi.DefaultValue : "nil");
+				ret += $" = {(pi.DefaultValue != null ? pi.DefaultValue : "nil")}";
```
Comment (pchote) [inline OpenRA.Game/Scripting/ScriptMemberExts.cs]: `pi.DefaultValue ?? "nil"`

File: OpenRA.Game/Server/Server.cs [hunk 4]
```diff
@@ -816,7 +816,7 @@ public void DispatchOrdersToClients(Connection conn, int frame, byte[] data)
 					if (data.Length == Order.SyncHashOrderLength)
 						HandleSyncOrder(frame, data);
 					else
-						Log.Write("server", "Dropped sync order with length {0} from client {1}. Expected length {2}.".F(data.Length, from, Order.SyncHashOrderLength));
+						Log.Write("server", $"Dropped sync order with length {data.Length} from client {@from}. Expected length {Order.SyncHashOrderLength}.");
```
Comment (pchote) [inline OpenRA.Game/Server/Server.cs]: Was the extra `@` here intentional? Rider says its not needed.

File: OpenRA.Mods.Cnc/Graphics/Voxel.cs [hunk 5]
```diff
@@ -60,9 +60,9 @@ public Voxel(VoxelLoader loader, VxlReader vxl, HvaReader hva, (string Vxl, stri
 		public float[] TransformationMatrix(uint limb, uint frame)
 		{
 			if (frame >= frames)
-				throw new ArgumentOutOfRangeException(nameof(frame), "Only {0} frames exist.".F(frames));
+				throw new ArgumentOutOfRangeException(nameof(frame), $"Only {frames} frames exist.");
 			if (limb >= limbs)
-				throw new ArgumentOutOfRangeException(nameof(limb), "Only {1} limbs exist.".F(limbs));
+				throw new ArgumentOutOfRangeException(nameof(limb), string.Format("Only {1} limbs exist.", limbs));
```
Comment (pchote) [inline OpenRA.Mods.Cnc/Graphics/Voxel.cs]: Bogus format string broke your script?

`$"Only {limbs} limbs exist."`.

File: OpenRA.Mods.Common/Util.cs [hunk 6]
```diff
@@ -212,19 +212,19 @@ public static int RandomDelay(World world, int[] range)
 		public static string FriendlyTypeName(Type t)
 		{
 			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(HashSet<>))
-				return "Set of {0}".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
+				return string.Format("Set of {0}", t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: Convert to interpolated string?

File: OpenRA.Mods.Common/Util.cs [hunk 7]
```diff
@@ -212,19 +212,19 @@ public static int RandomDelay(World world, int[] range)
...
 
 			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(Dictionary<,>))
-				return "Mapping of {0} to {1}".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
+				return string.Format("Mapping of {0} to {1}", t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: Bogus format string broke your script?

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs [hunk 8]
```diff
@@ -510,14 +510,14 @@ public static void SetupKickSpectatorsWidget(Widget parent, OrderManager orderMa
 				{
 					Game.LoadWidget(null, "KICK_SPECTATORS_DIALOG", lobby.Get("TOP_PANELS_ROOT"), new WidgetArgs
 					{
-						{ "clientCount", "{0}".F(spectatorCount) },
+						{ "clientCount", $"{spectatorCount}" },
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs]: Redundant format string?

**Review by pchote — COMMENTED** [review]
I still think its really awful when string interpolations include arbitrarily complex expressions, but I don't think it is productive to get hung up over that here. The majority of cases in this PR are simple variables and are a clear improvement, and the ugly cases can be cleaned up the next time someone touches the impacted code.

LGTM except for a couple of small nits:

File: OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs (hunk 8 above)
Comment (pchote) [reply OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs]: We probably want this to be an int, and then change `string clientCount` in `KickSpectatorsLogic` to match.

File: OpenRA.Mods.Common/Util.cs (hunk 7 above)
Comment (teinarss) [reply OpenRA.Mods.Common/Util.cs]: Looks like it, do we want a string with the key and value?

Comment (pchote) [reply OpenRA.Mods.Common/Util.cs]: I think so yeah.

**Comment by teinarss** [comment]
Updated.

File: OpenRA.Game/Game.cs [hunk 9]
```diff
@@ -1,4 +1,4 @@
-#region Copyright & License Information
+	#region Copyright & License Information
```
Comment (pchote) [inline OpenRA.Game/Game.cs]: This doesn't look intentional

**Comment by teinarss** [comment]
Fixed.

**Review by pchote — DISMISSED** [review]

**Comment by reaperrr** [comment]
Needs rebase, sorry. I guess we should merge this before any other PRs it might conflict with.

**Comment by teinarss** [comment]
Updated.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/13f89d35b2a38bfeb3a053e207a18fbbc9aedace...ce3669b0e1413527e261f5c820aa3b968eb659df

**Pull request merged**